R/estimate_contrasts.R

```diff
@@ -11,9 +11,49 @@
 #'   "bonferroni", "BH", "BY", "fdr" or "none". See the p-value adjustment
 #'   section in the `emmeans::test` documentation.
 #' @param adjust Deprecated in favour of `p_adjust`.
+#' @param effectsize Desired measure of standardized effect size, one of "none"
+#' (default), "emmeans", "marginal", or "bootES".
+#' @param bootES_type Specifies the type of effect-size measure to
+#' estimate when using `effectsize = "bootES"`. One of `c("unstandardized",
+#' "cohens.d", "hedges.g", "cohens.d.sigma", "r", "akp.robust.d")`. See`
+#' effect.type` argument of [bootES::bootES] for details.
+#' @param bootstraps The number of bootstrap resamples to perform.
 #'
 #' @inherit estimate_slopes details
 #'
+#' @section Effect Size: By default, `estimate_contrasts` reports no
+#' standardized effect size on purpose. Should one request one, some things
+#' are to keep in mind. As the authors of `emmeans` write, "There is
+#' substantial disagreement among practitioners on what is the appropriate
+#' sigma to use in computing effect sizes; or, indeed, whether any effect-size
+#' measure is appropriate for some situations. The user is completely
+#' responsible for specifying appropriate parameters (or for failing to do
+#' so)."
+#'
+#' In particular, effect size method `"bootES"` does not correct
+#' for covariates in the model, so should probably only be used when there is
+#' just one categorical predictor (with however many levels). Some believe that
+#' if there are multiple predictors or any covariates, it is important to
+#' re-compute sigma adding back in the response variance associated with the
+#' variables that aren't part of the contrast.
+#'
+#' `effectsize = "emmeans"` uses [emmeans::eff_size] with
+#' `sigma = stats::sigma(model)`, `edf = stats::df.residual(model)` and
+#' `method = "identity")`. This standardizes using the MSE (sigma). Some believe
+#' this works when the contrasts are the only predictors in the model, but not
+#' when there are covariates. The response variance accounted for by the
+#' covariates should not be removed from the SD used to standardize. Otherwise,
+#' _d_ will be overestimated.
+#'
+#' `effectsize = "marginal"` uses the following formula to compute effect
+#' size: `d_adj <- t * se_b / sigma * sqrt(1 - R2_cov)`. This standardized
```

Comment:
Note that `t * se_b` is equal to the raw difference - so this can be rewritten as `difference * (1- R2)/ sigma`, which I think is more clear.

Reply:
This also means that `emmeans::eff_size()` can be used here with `sigma = stats::sigma(model) / (1 - R2_cov)` (and some reasonable `edf`)

Reply:
Ok, one thing to note, is that previously I hardcoded sigma as as `sigma = stats::sigma(model)`. Would it be preferable to add a new argument `sigma` with that as the default, but then people can specify a different one when using the `emmeans` method? And Should this argument also apply to Brenton's `marginal` method? (again, instead of hard coding it as `sigma(model)`)

Reply:
Similarly, in the `emmeans` method, `edf` is currently hardcoded as `edf = stats::df.residual(model)`, and method  as `method = "identity"`. But I can add arguments for those as well (I'm always unsure whether the ellipsis three dots would work when they can be used in several functions with similar argument names, it would probably create conflict for `method` for example).

Reply:
Well, in that case you're just making a wrapper around `eff_size`, yeah?

Reply:
Yes. But I don't know what we should do (wrapper around `eff_size` with some defaults, or fixed values). Do you have a preference?

Reply:
nothing 😜🤷‍♂️